### PR TITLE
feat: allow overriding readOnly behavior of dateEditor

### PR DESCRIPTION
**Pull request by zewa666**
So this is a general idea of how I thought of tackling the realted issue https://github.com/ghiscoding/slickgrid-universal/issues/1684

No tests or docs yet as I first would like to validate what you think about the feature request @ghiscoding

### CODE REVIEW

File: packages/common/src/editors/dateEditor.ts [hunk 1]
```diff
@@ -186,7 +186,7 @@ export class DateEditor implements Editor {
           title: this.columnEditor && this.columnEditor.title || '',
           className: inputCssClasses.replace(/\./g, ' '),
           dataset: { input: '', defaultdate: this.defaultDate },
-          readOnly: this.columnEditor.readOnly === false ? false : true,
+          readOnly: this.columnEditor.editorOptions?.allowEdit === true ? true : false,
```
Comment (ghiscoding) [inline packages/common/src/editors/dateEditor.ts:189]: probably easier to do `!!(this.columnEditor.editorOptions?.allowEdit)`

Comment (zewa666) [reply packages/common/src/editors/dateEditor.ts:189]: passing on options to other options I'm typically careful since true is not the same as truthy but you're probably right here

**Comment by zewa666** [comment]
need to update the docs and e2e, but first I need to get the kids to sleep ;)

File: packages/common/src/editors/dateEditor.ts (hunk 1 above)
Comment (ghiscoding) [reply packages/common/src/editors/dateEditor.ts:189]: I've used this `!!(var)` for years, it's an easy shortcut to convert the output to a boolean

Comment (zewa666) [reply packages/common/src/editors/dateEditor.ts:189]: thats right but `"false" == true`  and can happen relatively easy as consumer of the grid. chances are the user meant false just mistankely added quotes

Comment (ghiscoding) [reply packages/common/src/editors/dateEditor.ts:189]: ahhh if that happens then yes I agree 😆

File: test/cypress/e2e/example11.cy.ts [hunk 2]
```diff
@@ -22,6 +22,20 @@ describe('Example 11 - Batch Editing', () => {
     cy.get('h3').should('contain', 'Example 11 - Batch Editing');
   });
 
+  it('should input values directly in input of datepicker', () => {
+    cy.get(`[style="top: ${GRID_ROW_HEIGHT * 0}px;"] > .slick-cell:nth(5)`)
+      .click();
+
+    cy.get(`.input-group-editor`)
+      .focus()
+      .type('{backspace}'.repeat(10))
```
Comment (ghiscoding) [inline test/cypress/e2e/example11.cy.ts:31]: might be easier to simply use Ctrl+Backspace

File: examples/vite-demo-vanilla-bundle/src/examples/example11.ts [hunk 3]
```diff
@@ -175,7 +175,7 @@ export default class Example11 {
         type: FieldType.date, outputType: FieldType.dateIso,
         filterable: true,
         filter: { model: Filters.compoundDate },
-        editor: { model: Editors.date, massUpdate: true },
+        editor: { model: Editors.date, massUpdate: true, disabled: false, editorOptions: { allowInput: true } },
```
Comment (ghiscoding) [inline examples/vite-demo-vanilla-bundle/src/examples/example11.ts:178]: is there a particular reason to add the `disabled: false`?

Comment (zewa666) [reply examples/vite-demo-vanilla-bundle/src/examples/example11.ts:178]: just an oversite from a manual test. good catch

File: test/cypress/e2e/example11.cy.ts (hunk 2 above)
Comment (zewa666) [reply test/cypress/e2e/example11.cy.ts:31]: {ctrl}{backspace} didn't work for me. But I doubt there is any con with this approach. So if you're fine I'd just keep it that way

Comment (ghiscoding) [reply test/cypress/e2e/example11.cy.ts:31]: ok sure whatever works is fine, just thought the alternative was shorter but if it doesn't work then too bad

Comment (ghiscoding) [reply test/cypress/e2e/example11.cy.ts:31]: the E2E test still fails though, should we use `{enter}` key after typing the date?

**Comment by zewa666** [comment]
hmm not sure why the tests now failed ... is this some random sideeffect?

**Comment by ghiscoding** [comment]
> hmm not sure why the tests now failed ... is this some random sideeffect?

I don't think so, I saw it failing at least twice in your PR, so I think it's a legitimate failure. Does it pass locally?

File: packages/common/src/editors/dateEditor.ts (hunk 1 above)
Comment (ghiscoding) [reply packages/common/src/editors/dateEditor.ts:189]: @zewa666 ah this is probably why it fails, it looks like you forgot rename this readonly to `allowInput`

**Comment by zewa666** [comment]
yes it did, and the above mentioned readOnly is fine, thats the attribute that gets attached to the input.

I'll take a look again later today

**Comment by zewa666** [comment]
alright the issue was as expected, those other tests looked for the number of modified cells, which got messed up due to my test starting at the begin and adding another one. Moved it to the end and now all should be fine

**Comment by ghiscoding** [comment]
> alright the issue was as expected, those other tests looked for the number of modified cells, which got messed up due to my test starting at the begin and adding another one. Moved it to the end and now all should be fine

ah yeah that's the side effect of configuring Cypress like I did where the sequence is important. That is actually discouraged by Cypress, but I prefer to use this approach to avoid having tests that are super long because the Cypress default approach is to make every tests independent and I don't like that part so..

@zewa666  On the a different subject, I had a user who opened an issue in Angular-Slickgrid complaining that date sorting was slow for 50k rows and I closed the issue because I thought I couldn't improve it further and also because the person is using an old version... however, most of my examples are using Date object which is much quicker, but when we use another format (i.e. US format) then the date has to be parsed and then convert to timestamp to sort and that is in fact slowing down a lot the sort (25k rows took a few seconds to sort). The bad thing is that the default browser sort will probably read the same row multiple times, meaning parsing the same date multiple times. So my question is have you ever use any other Sort than the default browser `.sort()`? From what I found so far, some users provide suggestions but it's usually an entire new sort function which is not connectable to the default browser sort and that is probably not going to work with my project. Any suggestions? I'll search some more but ideally I would like to override the default browser sort, which I'm already doing, but use a different type of sort for large dataset (quick sort, or whatever else)

**Comment by zewa666** [comment]
closed the issue, sry for wrong tagging.

as for the sorting performance, since I'm using server side backends I cant really tell. what I can tell is that sorting more than 50k rows on the client sounds like a huge UX issue. really if its more than ~100 rows I highly question the use of client side at all. thats where the server excels

**Comment by ghiscoding** [comment]
yeah well you know, there's always a ton of users who still prefer to have everything loaded locally for whatever reasons (even if there's not valid reasons). I even saw 1 guy mentioning in SlickGrid repo that he loads few million rows which seems overkill to me, but if SlickGrid makes it usable then yeah you'll have users that go for it... 

I think that there's couple of reasons why pagination is not always good to use, there are some features that aren't supported with Pagination and those are Tree Data, Row Detail and most notably Grouping and this later one is probably a huge one for some

Anyway, I'll search some more and see if I can find anything that can improve perf, it's mostly just the Dates that are slow, sorting string or numbers are typically quite fast but Dates requiring parsing which slows it down a lot.

**Comment by zewa666** [comment]
could you perhaps create a sample repro to look at? I dig its about changing to a custom sorter but hard to follow without concrete examples

**Comment by ghiscoding** [comment]
Yeah I modified Example 2 with US Dates and that's when I noticed the much slower sort after I click on the Add 25K rows button and then try to sort by Start/Finish dates. I can open a draft PR with that after I update the Cypress tests that will surely fail, I'll do that later today after work hours :)
